File: core-py/src/openglass_core/attest/attest.py
from __future__ import annotations

import secrets
import time
from datetime import datetime, timezone
from typing import Any, Callable, Optional, TypeVar

import httpx
from openglass import AgentIdentity, Signature, base64url_encode, canonicalize, canonicalize_to_bytes, sha256, sig_input, sign_ed25519, to_hex

from ..policy.evaluate import evaluate
from ..policy.types import Policy, PolicyEvent, PolicyResult, RiskLevel
from .http_sign import signed_request
from .types import AttestationCloseStatement, AttestationOpen, AttestOptions, AttestResult

_RISK_ORDER = {"low": 0, "medium": 1, "high": 2}
_DEFAULT_BASE_URL = "https://openglass.glass"
_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"  # Crockford base32, no I/L/O/U

T = TypeVar("T")


def _random_ulid() -> str:
    return "".join(_ALPHABET[b % 32] for b in secrets.token_bytes(26))


def _now_iso() -> str:
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


def _with_retry(fn: Callable[[], T], attempts: int) -> T:
    last_err: Optional[BaseException] = None
    for i in range(attempts):
        try:
            return fn()
        except Exception as err:  # noqa: BLE001 - deliberately broad: any transport failure is retried
            last_err = err
            if i < attempts - 1:
                time.sleep(0.2 * (2**i))
    assert last_err is not None
    raise last_err


def _sign_purpose(purpose: str, obj: Any, identity: AgentIdentity) -> Signature:
    digest = sha256(canonicalize(obj).encode("utf-8"))
    return {"alg": "Ed25519", "kid": identity.kid, "sig": base64url_encode(sign_ed25519(sig_input(purpose, digest), identity.private_key))}


def _default_on_error(error: Exception) -> None:
    print(f"[openglass-core] attestation failed, continuing without one: {error}")


def evaluate_and_attest(identity: AgentIdentity, policy: Policy, event: PolicyEvent, opts: Optional[AttestOptions] = None) -> AttestResult:
    """The one primitive every integration calls. See core-js's `attest.ts` for the full
    design rationale (fail-open, retry, one attestation per risky event) — this is its
    Python port, synchronous to match `openglass-sdk`'s own client."""
    o: AttestOptions = opts or {}
    verdict: PolicyResult = evaluate(policy, event)
    min_risk: RiskLevel = o.get("min_risk", "high")
    if _RISK_ORDER[verdict["risk"]] < _RISK_ORDER[min_risk]:
        return {"attested": False, "reason": "below_threshold", "risk": verdict["risk"], "matches": verdict["matches"]}

    base_url = o.get("base_url", _DEFAULT_BASE_URL)
    mode = o.get("mode", "notary")
    retries = o.get("retries", 3)
    client: Optional[httpx.Client] = o.get("http_client")
    on_error = o.get("on_error", _default_on_error)

    try:
        attestation_id = f"att_{_random_ulid()}"
        open_stmt: AttestationOpen = {
            "v": 1,
            "type": "openglass.attestation_open",
            "attestationId": attestation_id,
            "mode": mode,
            "purpose": ", ".join(m["id"] for m in verdict["matches"]) or "openglass-policy match",
            "attestor": {"agentId": identity.agent_id or "", "kid": identity.kid, "publicKey": base64url_encode(identity.public_key)},
            "createdAt": _now_iso(),
        }
        open_signature = _sign_purpose("attestation_open", open_stmt, identity)
        opened = _with_retry(
            lambda: signed_request(base_url, "POST", "/v1/attestations", {"open": open_stmt, "openSignature": open_signature, "idleTimeoutSec": o.get("idle_timeout_sec")}, identity, client),
            retries,
        )
        genesis_hash: str = opened["attestation"]["genesisHash"]

        verdict_payload = {"event": event, "verdict": verdict}
        payload = verdict_payload if mode == "relay" else None
        payload_hash = to_hex(sha256(canonicalize_to_bytes(verdict_payload)))
        envelope = {
            "v": 1,
            "type": "openglass.message",
            "sessionId": attestation_id,
            "seq": 1,
            "prevHash": genesis_hash,
            "sender": {"agentId": identity.agent_id, "kid": identity.kid},
            "contentType": "application/json",
            "payloadHash": payload_hash,
            "sentAt": _now_iso(),
        }
        hash_bytes = sha256(bytes.fromhex(genesis_hash) + canonicalize_to_bytes(envelope))
        event_hash = to_hex(hash_bytes)
        signature: Signature = {"alg": "Ed25519", "kid": identity.kid, "sig": base64url_encode(sign_ed25519(sig_input("message", hash_bytes), identity.private_key))}
        appended = _with_retry(
            lambda: signed_request(base_url, "POST", f"/v1/attestations/{attestation_id}/events", {"envelope": envelope, "hash": event_hash, "signature": signature, "payload": payload}, identity, client),
            retries,
        )
        head = appended["head"]

        statement: AttestationCloseStatement = {"v": 1, "type": "openglass.close", "sessionId": attestation_id, "headSeq": head["seq"], "headHash": head["hash"], "closedAt": _now_iso()}
        close_signature = _sign_purpose("close", statement, identity)
        _with_retry(lambda: signed_request(base_url, "POST", f"/v1/attestations/{attestation_id}/close", {"statement": statement, "signature": close_signature}, identity, client), retries)

        return {"attested": True, "attestationId": attestation_id, "risk": verdict["risk"], "matches": verdict["matches"]}
    except Exception as error:  # noqa: BLE001 - fail open: never propagate, always report
        on_error(error)
        return {"attested": False, "reason": "api_unreachable", "risk": verdict["risk"], "matches": verdict["matches"], "error": error}


def witness_if_risky(
    identity: AgentIdentity,
    policy: Policy,
    event: PolicyEvent,
    action: Callable[[], T],
    opts: Optional[AttestOptions] = None,
) -> tuple[T, AttestResult]:
    """Attests first (so the record exists even if `action` later raises), then runs
    `action`. See core-js's `witnessIfRisky` for the same rationale."""
    attestation = evaluate_and_attest(identity, policy, event, opts)
    result = action()
    return result, attestation

File: core-js/src/attest/attest.ts
import { base64UrlEncode, canonicalize, canonicalizeToBytes, hex, sha256, sigInput, signEd25519, type AgentIdentity, type Signature } from "openglass-sdk";
import { evaluate } from "../policy/evaluate.js";
import type { Policy, PolicyEvent, RiskLevel } from "../policy/types.js";
import { signedRequest } from "./httpSign.js";
import type { AttestApiView, AttestationCloseStatement, AttestationOpen, AttestOptions, AttestResult } from "./types.js";
import { randomUlid } from "./ulid.js";

const RISK_ORDER: Record<RiskLevel, number> = { low: 0, medium: 1, high: 2 };
const DEFAULT_BASE_URL = "https://openglass.glass";

async function withRetry<T>(fn: () => Promise<T>, attempts: number): Promise<T> {
  let lastErr: unknown;
  for (let i = 0; i < attempts; i++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      if (i < attempts - 1) await new Promise((r) => setTimeout(r, 200 * 2 ** i));
    }
  }
  throw lastErr;
}

function signPurpose<T>(purpose: string, obj: T, identity: AgentIdentity): Signature {
  const digest = sha256(Buffer.from(canonicalize(obj), "utf8"));
  return { alg: "Ed25519", kid: identity.kid, sig: base64UrlEncode(signEd25519(sigInput(purpose, digest), identity.privateKey)) };
}

/**
 * The one primitive every integration calls: evaluate `event` against `policy`, and if
 * the risk meets `opts.minRisk` (default "high"), open a one-party attestation (SPEC
 * §12) recording the event and its verdict, then close it immediately — a single
 * witnessed record per risky action, requiring your own agent identity to already be
 * registered and claimed.
 *
 * Fails open by design: a network problem or an OpenGlass outage is retried
 * (`opts.retries`, default 3, exponential backoff) and then reported via `opts.onError`
 * — it never throws, so it never blocks whatever real action this event is describing.
 * That's the point of an integration kit: OpenGlass being down should never be the
 * reason your agent's actual work stops.
 */
export async function evaluateAndAttest(identity: AgentIdentity, policy: Policy, event: PolicyEvent, opts: AttestOptions = {}): Promise<AttestResult> {
  const verdict = evaluate(policy, event);
  const minRisk = opts.minRisk ?? "high";
  if (RISK_ORDER[verdict.risk] < RISK_ORDER[minRisk]) {
    return { attested: false, reason: "below_threshold", risk: verdict.risk, matches: verdict.matches };
  }

  const baseUrl = opts.baseUrl ?? DEFAULT_BASE_URL;
  const mode = opts.mode ?? "notary";
  const retries = opts.retries ?? 3;
  const fetchImpl = opts.fetchImpl ?? fetch;
  const onError = opts.onError ?? ((err: unknown) => console.error("[openglass-core] attestation failed, continuing without one:", err));

  try {
    const attestationId = `att_${randomUlid()}`;
    const open: AttestationOpen = {
      v: 1,
      type: "openglass.attestation_open",
      attestationId,
      mode,
      purpose: verdict.matches.map((m) => m.id).join(", ") || "openglass-policy match",
      attestor: { agentId: identity.agentId!, kid: identity.kid, publicKey: base64UrlEncode(identity.publicKey) },
      createdAt: new Date().toISOString(),
    };
    const openSignature = signPurpose("attestation_open", open, identity);
    const { attestation } = await withRetry(
      () => signedRequest<{ attestation: AttestApiView }>(baseUrl, "POST", "/v1/attestations", { open, openSignature, idleTimeoutSec: opts.idleTimeoutSec }, identity, fetchImpl),
      retries,
    );

    const verdictPayload = { event, verdict };
    const payload = mode === "relay" ? verdictPayload : undefined;
    const payloadHash = hex(sha256(canonicalizeToBytes(verdictPayload)));
    const envelope = {
      v: 1 as const,
      type: "openglass.message" as const,
      sessionId: attestationId,
      seq: 1,
      prevHash: attestation.genesisHash,
      sender: { agentId: identity.agentId!, kid: identity.kid },
      contentType: "application/json",
      payloadHash,
      sentAt: new Date().toISOString(),
    };
    const hashBytes = sha256(Buffer.concat([Buffer.from(attestation.genesisHash, "hex"), Buffer.from(canonicalize(envelope), "utf8")]));
    const hash = hex(hashBytes);
    const signature: Signature = { alg: "Ed25519", kid: identity.kid, sig: base64UrlEncode(signEd25519(sigInput("message", hashBytes), identity.privateKey)) };
    const { head } = await withRetry(
      () => signedRequest<{ head: { seq: number; hash: string } }>(baseUrl, "POST", `/v1/attestations/${attestationId}/events`, { envelope, hash, signature, payload }, identity, fetchImpl),
      retries,
    );

    const statement: AttestationCloseStatement = { v: 1, type: "openglass.close", sessionId: attestationId, headSeq: head.seq, headHash: head.hash, closedAt: new Date().toISOString() };
    const closeSignature = signPurpose("close", statement, identity);
    await withRetry(() => signedRequest(baseUrl, "POST", `/v1/attestations/${attestationId}/close`, { statement, signature: closeSignature }, identity, fetchImpl), retries);

    return { attested: true, attestationId, risk: verdict.risk, matches: verdict.matches };
  } catch (error) {
    onError(error);
    return { attested: false, reason: "api_unreachable", risk: verdict.risk, matches: verdict.matches, error };
  }
}

/**
 * Convenience wrapper for wrapping a callback directly (mirroring `openglass-sdk`'s own
 * `witness()`): attests first — so the record exists even if `action` later throws — then
 * runs `action`. Most integrations (openglass-otel included) observe events after the
 * fact and call `evaluateAndAttest` directly instead; this is for the case where you're
 * intercepting a call before it happens.
 */
export async function witnessIfRisky<T>(
  identity: AgentIdentity,
  policy: Policy,
  event: PolicyEvent,
  action: () => Promise<T> | T,
  opts: AttestOptions = {},
): Promise<{ result: T; attestation: AttestResult }> {
  const attestation = await evaluateAndAttest(identity, policy, event, opts);
  const result = await action();
  return { result, attestation };
}
